tests/std/tests/P0798R8_monadic_operations_for_std_optional/test.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+using namespace std;
+
+struct Immovable {
+    constexpr Immovable(int x) : v(x) {}
+    Immovable(const Immovable&) = delete;
+    Immovable(Immovable&&)      = delete;
+    Immovable& operator=(const Immovable&) = delete;
+    Immovable& operator=(Immovable&&) = delete;
+    constexpr ~Immovable() {}
+
+    int v;
+};
+
+struct Thingy {
+    optional<int> x;
+    constexpr int member_func() const {
+        return 22;
+    }
+};
+
+template <class T>
+struct fn {
+    template <class U>
+    constexpr auto operator()(U&&) & {
+        static_assert(is_same_v<T, U&&>);
+        return optional<int>{33};
+    }
+
+    template <class U>
+    constexpr auto operator()(U&&) && {
+        static_assert(is_same_v<T, U&&>);
+        return Immovable{44};
+    }
+
+    constexpr auto operator()() && {
+        return optional<Thingy>{Thingy{55}};
+    }
+};
+
+template <class Optional>
+constexpr void test_impl(Optional&& nonempty, Optional&& empty_optional) {

```

Comment:
Should we assert the preconditions here, namely that `empty_optional` is empty and `nonempty` is not`?